// @lafjs/cloud@1.0.0/dist/cloud.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cloud = void 0;
const axios_1 = require("axios");
const storage_1 = require("./storage");
class Cloud {
    constructor() {
        /**
         * Sending an HTTP request is actually an Axios instance. You can refer to the Axios documentation directly.
         * @deprecated this is deprecated and will be removed in future, use the global `fetch()` directly @see https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
         * @see https://axios-http.com/docs/intro
         */
        this.fetch = axios_1.default;
        /**
         * Invoke cloud function
         * @deprecated Just import the cloud function directly, and then call it
         */
        this.invoke = (name, param) => {
            return this.cloud.invoke(name, param);
        };
        this.getToken = (param) => {
            return this.cloud.getToken(param);
        };
        this.parseToken = (token) => {
            return this.cloud.parseToken(token);
        };
        this.storage = new storage_1.CloudStorage();
    }
    get cloud() {
        if (globalThis.createCloudSdk && !Cloud.create) {
            Cloud.create = globalThis.createCloudSdk;
        }
        if (!this._cloud) {
            this._cloud = Cloud.create();
        }
        return this._cloud;
    }
    database() {
        return this.cloud.database();
    }
    get shared() {
        return this.cloud.shared;
    }
    get mongo() {
        return this.cloud.mongo;
    }
    get sockets() {
        return this.cloud.sockets;
    }
    get appid() {
        return this.cloud.appid;
    }
    /**
     * @deprecated this is deprecated and will be removed in future, use `process.env` instead
     */
    get env() {
        return this.cloud.env;
    }
}
exports.Cloud = Cloud;
