// @lafjs/cloud@1.0.0/dist/storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudStorageBucket = exports.CloudStorage = void 0;
const assert = require("node:assert");
const client_s3_1 = require("@aws-sdk/client-s3");
const s3_request_presigner_1 = require("@aws-sdk/s3-request-presigner");
const util_1 = require("./util");
/**
 * `ICloudStorage` is an interface for cloud storage services.
 */
class CloudStorage {
    get appid() {
        assert(process.env.APPID, 'APPID is required');
        return process.env.APPID;
    }
    get externalEndpoint() {
        assert(process.env.OSS_EXTERNAL_ENDPOINT, 'OSS_EXTERNAL_ENDPOINT is required');
        return process.env.OSS_EXTERNAL_ENDPOINT;
    }
    get internalEndpoint() {
        assert(process.env.OSS_INTERNAL_ENDPOINT, 'OSS_INTERNAL_ENDPOINT is required');
        return process.env.OSS_INTERNAL_ENDPOINT;
    }
    /**
     * Get external S3 client of `@aws-sdk/client-s3`.
     * You can use this client to access the bucket through the external endpoint.
     * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3/
     * @returns
     */
    getExternalS3Client() {
        if (!this._externalS3Client) {
            assert(process.env.OSS_EXTERNAL_ENDPOINT, 'OSS_EXTERNAL_ENDPOINT is required');
            assert(process.env.OSS_REGION, 'OSS_REGION is required');
            assert(process.env.OSS_ACCESS_KEY, 'OSS_ACCESS_KEY is required');
            assert(process.env.OSS_ACCESS_SECRET, 'OSS_ACCESS_SECRET is required');
            this._externalS3Client = new client_s3_1.S3({
                endpoint: process.env.OSS_EXTERNAL_ENDPOINT,
                region: process.env.OSS_REGION,
                credentials: {
                    accessKeyId: process.env.OSS_ACCESS_KEY,
                    secretAccessKey: process.env.OSS_ACCESS_SECRET,
                },
                forcePathStyle: true,
            });
        }
        return this._externalS3Client;
    }
    /**
     * Get internal S3 client of `@aws-sdk/client-s3`.
     * You can use this client to access the bucket through the internal endpoint.
     * @see https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3/
     */
    getInternalS3Client() {
        if (!this._internalS3Client) {
            assert(process.env.OSS_INTERNAL_ENDPOINT, 'OSS_INTERNAL_ENDPOINT is required');
            assert(process.env.OSS_REGION, 'OSS_REGION is required');
            assert(process.env.OSS_ACCESS_KEY, 'OSS_ACCESS_KEY is required');
            assert(process.env.OSS_ACCESS_SECRET, 'OSS_ACCESS_SECRET is required');
            this._internalS3Client = new client_s3_1.S3({
                endpoint: process.env.OSS_INTERNAL_ENDPOINT,
                region: process.env.OSS_REGION,
                credentials: {
                    accessKeyId: process.env.OSS_ACCESS_KEY,
                    secretAccessKey: process.env.OSS_ACCESS_SECRET,
                },
                forcePathStyle: true,
            });
        }
        return this._internalS3Client;
    }
    /**
     * Get bucket by bucket name
     * @returns
     */
    bucket(bucketName) {
        assert(bucketName, 'bucketName is required');
        if (util_1.IS_SEALAF || bucketName.startsWith(`${this.appid}-`)) {
            return new CloudStorageBucket(this, bucketName);
        }
        const name = `${this.appid}-${bucketName}`;
        return new CloudStorageBucket(this, name);
    }
}
exports.CloudStorage = CloudStorage;
class CloudStorageBucket {
    constructor(storage, name) {
        assert(storage, 'storage is required');
        assert(name, 'name is required');
        this.storage = storage;
        this.name = name;
    }
    /**
     * Read file from bucket
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @param options
     * @returns
     * @see https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html
     */
    async readFile(filename, options) {
        assert(filename, 'filename is required');
        const internal = this.storage.getInternalS3Client();
        const args = Object.assign({ Bucket: this.name, Key: filename }, options);
        const res = await internal.getObject(args);
        return res;
    }
    /**
     * Write file to bucket
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @param body body is the content of the object, it can be a string, a Buffer, a Readable stream, or a Blob
     * @param options
     * @returns
     * @see https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html
     */
    async writeFile(filename, body, options) {
        assert(filename, 'key is required');
        assert(body, 'body is required');
        const external = this.storage.getInternalS3Client();
        const args = Object.assign({ Bucket: this.name, Key: filename, Body: body }, options);
        const res = await external.putObject(args);
        return res;
    }
    /**
     * Delete file from bucket
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @param options
     * @returns
     * @see https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObject.html
     */
    async deleteFile(filename, options) {
        assert(filename, 'filename is required');
        const external = this.storage.getInternalS3Client();
        const args = Object.assign({ Bucket: this.name, Key: filename }, options);
        const res = await external.deleteObject(args);
        return res;
    }
    /**
     * List files in bucket
     * @param prefix prefix is the key prefix of the object, it can contain subdirectories, e.g. `a/b/`
     * @param options
     * @returns
     * @see https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html
     */
    async listFiles(options) {
        const internal = this.storage.getInternalS3Client();
        const args = Object.assign({ Bucket: this.name }, options);
        const res = await internal.listObjects(args);
        return res;
    }
    /**
     * Get external url of the file.
     * You can ONLY use this url to access file from `readonly` bucket or `public` bucket.
     * Use `getDownloadUrl()` to get a signed url for `private` bucket.
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @returns
     */
    externalUrl(filename) {
        assert(filename, 'filename is required');
        return `${this.storage.externalEndpoint}/${this.name}/${filename}`;
    }
    /**
     * Get internal url of the file.
     * You can ONLY use this url to access file from `readonly` bucket or `public` bucket.
     * Use `getDownloadUrl()` to get a signed url for `private` bucket.
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @returns
     */
    internalUrl(filename) {
        assert(filename, 'filename is required');
        return `${this.storage.internalEndpoint}/${this.name}/${filename}`;
    }
    /**
     * Get upload url, you can use this url to upload file directly to bucket
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @param expiresIn expiresIn is the seconds of the signed url, default is `3600` seconds which is 1 hour
     * @param options
     * @returns
     */
    async getUploadUrl(filename, expiresIn = 3600, options) {
        assert(filename, 'filename is required');
        const external = this.storage.getExternalS3Client();
        const args = new client_s3_1.PutObjectCommand(Object.assign({ Bucket: this.name, Key: filename }, options));
        const url = await (0, s3_request_presigner_1.getSignedUrl)(external, args, { expiresIn: expiresIn });
        return url;
    }
    /**
     * Get download url, you can use this url to download file directly from bucket
     * @param filename filename is the key of the object, it can contain subdirectories, e.g. `a/b/c.txt`
     * @param expiresIn expiresIn is the seconds of the signed url, default is `3600` seconds which is 1 hour
     * @param options
     * @returns
     */
    async getDownloadUrl(filename, expiresIn = 3600, options) {
        assert(filename, 'filename is required');
        const external = this.storage.getExternalS3Client();
        const args = new client_s3_1.GetObjectCommand(Object.assign({ Bucket: this.name, Key: filename }, options));
        const url = await (0, s3_request_presigner_1.getSignedUrl)(external, args, { expiresIn: expiresIn });
        return url;
    }
}
exports.CloudStorageBucket = CloudStorageBucket;
