// database-proxy@1.0.0/dist/processor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Processor = exports.ProcessTypes = void 0;
const assert = __importStar(require("assert"));
var ProcessTypes;
(function (ProcessTypes) {
    ProcessTypes["Validator"] = "validator";
})(ProcessTypes || (exports.ProcessTypes = ProcessTypes = {}));
class Processor {
    constructor(name, handler, config, type = ProcessTypes.Validator) {
        assert.ok(handler instanceof Function, `${name}'s handler must be callable`);
        this.name = name;
        this.handler = handler;
        this.type = type;
        this.config = config;
    }
    async run(context) {
        return await this.handler.call(this, this.config, context);
    }
}
exports.Processor = Processor;
